tests/test_bern.py

```diff
@@ -0,0 +1,167 @@
+import numpy as np
+import pytest
+from firedrake import (DirichletBC, FacetNormal, Function, FunctionSpace,
+                       SpatialCoordinate, TestFunction, TestFunctions,
+                       UnitIntervalMesh, UnitSquareMesh, cos, diff, div, dot,
+                       ds, dx, errornorm, exp, grad, inner, norm, pi, project,
+                       split)
+from irksome import Dt, MeshConstant, RadauIIA, TimeStepper
+from ufl.algorithms import expand_derivatives
+
+lu_params = {
+    "snes_type": "ksponly",
+    "ksp_type": "preonly",
+    "mat_type": "aij",
+    "pc_type": "lu"
+}
+
+vi_params = {
+    "snes_type": "vinewtonrsls",
+    "snes_max_it": 300,
+    "snes_atol": 1.e-8,
+    "ksp_type": "preonly",
+    "pc_type": "lu",
+}
+
+
+def heat(n, deg, butcher_tableau, solver_parameters,
+         **kwargs):
+    N = 2**n
+    msh = UnitIntervalMesh(N)
+
+    V = FunctionSpace(msh, "Bernstein", deg)
+    x, = SpatialCoordinate(msh)
+
+    MC = MeshConstant(msh)
+    t = MC.Constant(0.0)
+    dt = MC.Constant(2.0 / N)
+
+    uexact = exp(-t) * cos(pi * x)**2
+    rhs = expand_derivatives(diff(uexact, t)) - div(grad(uexact))
+
+    u = project(uexact, V)
+
+    v = TestFunction(V)
+
+    F = (inner(Dt(u), v) * dx + inner(grad(u), grad(v)) * dx
+         - inner(rhs, v) * dx)
+
+    bc = DirichletBC(V, uexact, "on_boundary")
+
+    stepper = TimeStepper(F, butcher_tableau, t, dt, u,
+                          bcs=bc, solver_parameters=solver_parameters,
+                          **kwargs)
+
+    while (float(t) < 1.0):
+        if (float(t) + float(dt) > 1.0):
+            dt.assign(1.0 - float(t))
+        stepper.advance()
+        t.assign(float(t) + float(dt))
+
+    return errornorm(uexact, u) / norm(uexact)
+
+
+def mixed_heat(n, deg, butcher_tableau, solver_parameters,
+               **kwargs):
+    N = 2**n
+    msh = UnitSquareMesh(N, N)
+
+    V = FunctionSpace(msh, "RT", deg)
+    el_type = "Bernstein" if deg > 1 else "DG"
+    W = FunctionSpace(msh, el_type, deg-1)
+
+    Z = V * W
+
+    x, y = SpatialCoordinate(msh)
+
+    MC = MeshConstant(msh)
+    t = MC.Constant(0.0)
+    dt = MC.Constant(1.0 / N)
+
+    pexact = (cos(pi * x) * cos(pi * y))**2 * exp(-t)
+    uexact = -grad(pexact)
+
+    up = Function(Z)
+    u, p = split(up)
+
+    v, w = TestFunctions(Z)
+
+    n = FacetNormal(msh)
+
+    rhs = expand_derivatives(diff(pexact, t) + div(uexact))
+
+    F = (inner(Dt(p), w) * dx
+         + inner(div(u), w) * dx
+         - inner(rhs, w) * dx
+         + inner(u, v) * dx
+         - inner(p, div(v)) * dx
+         + inner(pexact, dot(v, n)) * ds)
+
+    up.subfunctions[1].project(pexact)
```

Comment:
Why is `up.subfunctions[0]` not initialized?

Reply:
I think strictly speaking it doesn't matter, this sets the initial pressure field and the solver will compute a velocity field in the next timestep. The comparison in lines 113-114 is just against the final value of the exact and numerical solutions; if the test instead used, say, the time integral of the discrepancy then you'd be testing things incorrectly by not setting `up.subfunctions[0]` as well.

Reply:
Right, there isn't an official initial condition for the flux variable for the PDE, but it doesn't hurt anything to set the initial flux also (I just checked).